LayerSwitcher_#561 improvement

martyrion opened the pull request:
New Layer Switcher:

- movable window
- option to create without a switcher (button)
- option to have the sections expanded
- minimize/maximize window
- expand/collapse sections
- vertical resize window
- double click header to reset original position

#### **Verify that...**

- [x] `npm test` passes
- [x] `npm run api` passes
- [x] `npm run lint` passes
- [x] `npm run build` still works

martyrion committed:
Alternative LayerSwitcher

martyrion committed:
Layer Switcher

martyrion committed:
Redesigned LayerSwitcher

New layer switcher, with expandable lists, movable window, etc

martyrion committed:
Whitespace

martyrion committed:
Delete LayerSwitcher old file

Zemledelec commented on css/og.css:
```diff
@@ -200,81 +221,178 @@
+.og-dialog-header-btn>div {
+    width: 24px;
+    height: 24px;
+    background-size: 14px !important;
```
Is it possible to remove !important?

martyrion replied:
I will check.

Zemledelec commented on src/og/control/LayerSwitcherv2.js:
```diff
@@ -0,0 +1,490 @@
+    }
+
+    // BASIC DATA COLLECTION-PREPARATION FUNCTIONS
+    planetDataBasic = () => {
```
No need  methods to be arrow functions

martyrion replied:
Good, I will change that

Zemledelec commented on src/og/control/LayerSwitcherv2.js:
```diff
@@ -0,0 +1,490 @@
+
+            // If I clicked elsewhere and the dialog is unpinned, then hide dialog and set menu btn to OFF
+            if (whereIclicked == 'outside' && D_header_pin.dataset.attachement == 'UNPINNED') {
+                D_dialog.classList.add('og-not-visible')
```
In the openglobus js is only camelCase style in the code. Rename variables, please.

martyrion replied:
Hey! I recall when you said maybe using $ to distinguish DOM elements from other variables. I used this convention. I will rename them though

Zemledelec commented on src/og/control/LayerSwitcherv2.js:
```diff
@@ -0,0 +1,490 @@
+            // If I clicked elsewhere and the dialog is unpinned, then hide dialog and set menu btn to OFF
+            if (whereIclicked == 'outside' && D_header_pin.dataset.attachement == 'UNPINNED') {
+                D_dialog.classList.add('og-not-visible')
+                D_menu_btn.classList.add('og-OFF')
```
HTMLElement variables must be started from $, for instance $dialogHeaderPin

martyrion replied:
Good.

Zemledelec commented on src/og/control/LayerSwitcherv2.js:
```diff
@@ -0,0 +1,490 @@
+                D_dialog.classList.add('og-not-visible')
+                D_menu_btn.classList.add('og-OFF')
+                // If Iclicked on button, toggle dialog and menu btn
+            } else if (whereIclicked == 'on-btn') {
```
use ===

martyrion replied:
OK

Zemledelec commented on src/og/control/LayerSwitcherv2.js:
```diff
@@ -0,0 +1,490 @@
+        }
+
+        // Document
+        document.addEventListener('click', (e) => whereClickHandler(e))
```
If the layer switcher is removed there will be a leack. It must be resolved with removeEventListener in control deactivation virtual method.

martyrion replied:
Yes, I will fix it

martyrion replied:
I (think) it is fixed. I added a button to toggle activate/deactivate the switcher. For some reason after deactivation, I have to activate/deactivate/activate to get it to work again. I will look into it!

Zemledelec commented on src/og/control/LayerSwitcherv2.js:
```diff
@@ -0,0 +1,490 @@
+        }
+
+        const visibility = (object, nameConcat) => {
+            if (nameConcat == 'TerrainProviders' && planet.terrain == object) {
```
maybe it is better to refactor these cases into one?

martyrion replied:
Yes, noticed it also my self

Zemledelec commented on src/og/utils/functionComposition.js:
```diff
@@ -0,0 +1,28 @@
+
+// Example
+
+// const lower = str => str.toLowerCase();
```
clean up commented code

martyrion committed:
Performed required changes

- Removed !important from CSS
- Replace arrow with normal functions
- Refactored if...else
- Changed equalities to triple ===
- Remove listener on switcher deactivate
- Changed DOM variables to $camelCase
- Minor changes to layerSwitcher.html

Zemledelec reviewed (approved):
Good @martyrion ! In advance, lets refactor huge methods. It's good practice making functions as small as possible.
